### PR TITLE
Refactor `GradleDependencies`

### PR DESCRIPTION
**Pull request by ursachec**
- fix Exception in Gradle dependency downloading for Android app Gradle configs with duplicates
- test Gradle dependency downloading for sample Android app
- expose two new parameters for Gradle dependency downloading
- debug-log a lot more exceptions around Gradle dependency downloading
- introduce parameters to `DependencyResolver`
- load Gradle init script from temporary file instead of creating init script into GRADLE_HOME

### CODE REVIEW

File: joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/Kotlin2Cpg.scala [hunk 1]
```diff
@@ -51,9 +57,17 @@ class Kotlin2Cpg extends X2CpgFrontend[Config] {
...
           if (config.downloadDependencies) {
-            val paths = DependencyResolver.getDependencies(Paths.get(sourceDir))
+            val gradleParams = mutable.Map[GradleConfigKeys.GradleConfigKey, String]()
+            if (config.gradleProjectName.nonEmpty) {
+              gradleParams(GradleConfigKeys.ProjectName) = config.gradleProjectName.get
+            }
+            if (config.gradleConfigurationName.nonEmpty) {
+              gradleParams(GradleConfigKeys.ConfigurationName) = config.gradleConfigurationName.get
+            }
+            val resolverParams = DependencyResolverParams(Map(), gradleParams.toMap)
```
Comment (mpollmeier) [inline joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/Kotlin2Cpg.scala:69]:
```suggestion
            val gradleParams = Map(
              GradleConfigKeys.ProjectName -> config.gradleProjectName, 
              GradleConfigKeys.ConfigurationName -> config.gradleConfigurationName
            ).collect { case (key, Some(value)) => (key, value) }
            
            val resolverParams = DependencyResolverParams(Map.empty, gradleParams)
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/DependencyResolver.scala [hunk 2]
```diff
@@ -1,15 +1,46 @@
...
+
+  def isMavenBuild(codeDir: Path): Boolean = {
+    Files.exists(codeDir.resolve("pom.xml"))
+  }
+
+  def isGradleBuild(codeDir: Path): Boolean = {
+    Files
+      .walk(codeDir)
+      .filter(file => file.toString.endsWith(".gradle") || file.toString.endsWith(".gradle.kts"))
+      .count > 0
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/DependencyResolver.scala:44]:
```suggestion
      .anyMatch(file => file.toString.endsWith(".gradle") || file.toString.endsWith(".gradle.kts"))
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 3]
```diff
@@ -201,31 +188,30 @@ object GradleDependencies {
...
-          // .setStandardError(System.err)  // uncomment for debugging
+          .forTasks(initScript.taskName)
+          .withArguments("--init-script", initScriptFile.pathAsString)
+          .setStandardOutput(stdoutStream)
+          .setStandardError(stderrStream)
           .run()
       } catch {
         case t: Throwable =>
-          logger.warn(s"Caught exception while executing Gradle task: '${t.getMessage}'.")
+          logger.warn(s"Caught exception while executing Gradle task: '${t.getMessage}'")
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:205]: maybe you wanna pass the exception to the logger as a second param as well? maybe not... just saying

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 4]
```diff
@@ -201,31 +188,30 @@ object GradleDependencies {
...
+          logger.debug(s"FULL_EXCEPTION: \n", t)
       } finally {
+        logger.debug(s"Gradle task execution stdout: \n${stdoutStream.toString}")
+        logger.debug(s"Gradle task execution stderr: \n${stderrStream.toString}")
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:209]:
```suggestion
        logger.debug(s"Gradle task execution stdout: \n$stdoutStream")
        logger.debug(s"Gradle task execution stderr: \n$stderrStream")
```

**Review by mpollmeier — APPROVED** [review]

**Comment by ursachec** [comment]
@mpollmeier Fuer Ihren Beitrag bedank ich mich herzlich

**Pull request merged**